src/utp_stream.cpp

```diff
@@ -210,6 +210,152 @@ struct packet
 	std::uint8_t buf[1];
 };
 
+// single thread packet allocation packet pool
+// can handle 3 common cases of packet allocation size
+struct packet_pool
+{
+	packet *acquire(std::uint16_t allocate)
+	{
+#if TORRENT_USE_ASSERTS
+		check_thread();
+#endif
+		packet *p{ nullptr };
+		std::uint16_t should_allocate{ allocate };
+
+		if (allocate <= utp_header_size)
+		{
+			p = m_syn_stash.pop();
+			should_allocate = utp_header_size;
+		}
+		else if (allocate <= mtu_floor_size)
+		{
+			p = m_mtu_floor_stash.pop();
+			should_allocate = mtu_floor_size;
+		}
+		else if (allocate <= mtu_ceiling_size)
+		{
+			p = m_mtu_ceiling_stash.pop();
+			should_allocate = mtu_ceiling_size;
+		}
+
+		if (p == nullptr)
+			p = static_cast<packet*>(malloc(sizeof(packet) + should_allocate));
+
+		p->allocated = allocate;
+		return p;
+	}
+
+	void release(packet *p)
+	{
+#if TORRENT_USE_ASSERTS
+		check_thread();
+#endif
+		if (p == nullptr) return;
+		if (p->allocated <= utp_header_size)
+		{
+			if (m_syn_stash.push(p, utp_header_limit)) return;
+		}
+		else if (p->allocated <= mtu_floor_size)
+		{
+			if (m_mtu_floor_stash.push(p, mtu_floor_limit)) return;
+		}
+		else if (p->allocated <= mtu_ceiling_size)
+		{
+			if (m_mtu_ceiling_stash.push(p, mtu_ceiling_limit)) return;
+		}
+
+		free(p);
+	}
+
+	void flush()
+	{
+		m_syn_stash.flush();
+		m_mtu_floor_stash.flush();
+		m_mtu_ceiling_stash.flush();
+	}
+private:
+	struct packet_stash_item
+	{
+		packet *next;
+		std::size_t index;
+	};
+
+	struct packet_stash
+	{
+		packet *top{ nullptr };
+
+		~packet_stash()
+		{
+			flush();
+		}
+
+		void flush()
+		{
+			packet *p = top;
+			while (p != nullptr)
+			{
+				packet_stash_item *pi = reinterpret_cast<packet_stash_item *>(p);
+				packet* next = pi->next;
+				free(p);
+				p = next;
+			}
+		}
+
+		bool push(packet *p, std::size_t limit)
+		{
+			packet_stash_item *tpi = reinterpret_cast<packet_stash_item *>(top);
+			std::size_t top_index = (tpi == nullptr) ? 0 : tpi->index;
+			if (limit < top_index) return false;
+
+			packet_stash_item *pi = reinterpret_cast<packet_stash_item *>(p);
```

Comment:
Yes, there are a couple of places where libtorrent invokes undefined behavior using reinterpret_cast. Some are sort-of ok, like `entry`, because they use placement new and are careful to make sure alignment requirements are met. You still shouldn't be doing it, if you want to re-use storage like this you need to use a union type.